### DISCUSSION TITLE
understanding PGN 126208 (3 - Read Fields group)

### DISCUSSION THREAD

**Discussion by aldas** [canboat/canboat#344] [Q&A] [answered]
Canboat has concept of repeating fields. Up to two sets of repeating fields. Only PGN to have second set is some variations of 126208

NMEA site has some documentation about PGN 126208 variations  https://www.nmea.org/Assets/20140109%20nmea-2000-corrigendum-tc201401031%20pgn%20126208.pdf on page 9 is info for "NMEA - Read Fields - group function" variant (this is case when first field "Complex Request Group Function Code" = 0x03)

What I actually do not understand what that second field actually is and is it related to first set? 

So first set has its counter field as `7 Number of Selection Pairs` this tells us how many pairs of  `9 Field Number of First Selection Pair` + `10 Field Value of First Selection Pair` will be in this message.  

Also there is second set of fields. Their count is in this field `8 Number of Parameter Pairs to be Read`  which tells us how many times `13 Field Number of First Parameter Pair to be Read` field will be repeated

What I do not understand what these fields sets do? 

---------------------
Theory A:

First set of fields is list of pairs  (field index  + value) - I assume this is used to write with this message.  Second set is list of single field  (field index) - is it used for to instruct which field values to return?

Is this conceptually WRITE+READ command. Something similar to Modbus "FC23 - Read / Write Multiple Registers"  packet?

---------------
Probably this is correct

Theory B:

or `selection` and `parameter` refer to case when there could be multiple instances of parameters. ala you have multiple waypoints and want to read Nth waypoint field X value.  Or for example PGN 129540 (GNSS Sats in View) could have multiple sets of satellites and you want to read 2nd satellite in view field 10 (status) value?   

In that case - why `selection` set is pair of values and `parameter` set is single value?

So this is example of PGN 129540 (GNSS Sats in View) - how would select and parameter map to this?
```json
{
    "timestamp": "2022-02-11T13:26:34.151Z",
    "prio": 6,
    "src": 18,
    "dst": 255,
    "pgn": 129540,
    "description": "GNSS Sats in View",
    "fields": {
        "SID": {
            "value": 0,
            "bytes": "00"
        },
        "Range Residual Mode": {
            "value": null,
            "bytes": "03",
            "bits": "11"
        },
        "Sats in View": {
            "value": 2,
            "bytes": "02"
        },
        "list": [
            {
                "PRN": {
                    "value": 6,
                    "bytes": "06"
                },
                "Elevation": {
                    "value": 0.6283,
                    "bytes": "8B 18"
                },
                "Azimuth": {
                    "value": 1.3439,
                    "bytes": "7F 34"
                },
                "SNR": {
                    "value": 26.0,
                    "bytes": "28 0A"
                },
                "Range residuals": {
                    "value": null,
                    "bytes": "FF FF FF 7F"
                },
                "Status": {
                    "value": "Used",
                    "bytes": "02",
                    "bits": "0010"
                }
            },
            {
                "PRN": {
                    "value": 2,
                    "bytes": "02"
                },
                "Elevation": {
                    "value": 0.5061,
                    "bytes": "C5 13"
                },
                "Azimuth": {
                    "value": 2.3038,
                    "bytes": "FE 59"
                },
                "SNR": {
                    "value": 26.0,
                    "bytes": "28 0A"
                },
                "Range residuals": {
                    "value": null,
                    "bytes": "FF FF FF 7F"
                },
                "Status": {
                    "value": "Used",
                    "bytes": "02",
                    "bits": "0010"
                }
            }
        ]
    }
}
```

------------------

Does anybody have sample for real world packet?

p.s. there are no stupid questions, only stupid people :)

**Comment by keesverruijt** [comment, marked as answer]
It is _sort of_ like sample B.

The actual use case is where there is a device that has multiple sets of state. The parameter list is used to indicate which fields it should produce in the PGN. The selection pairs «ask for » a particular set of values, those where the selection pair parameters have the requested values.

For instance lets say you have a wind sensor that is reporting both an Apparent and a True wind direction. And theoretically it is « broken » and never reports them so you have to interrogate using PGN 126208. In that case you would send a PGN 126208 with a single selection pair (field 4, value 2) and a parameter list consisting of 2 parameters: (field 2, field 3).

In human terms this means: send me your values for field 2 (wind speed) and field 3 (wind angle) for the set of values where field 4 = 2 = Apparent. 

The above is obviously not needed as wind sensors tend to produce this data by themselves. But hopefully it explains it a bit.

Another example is a digital switch system that only reports PGN 127501 (Binary Switch Bank Status) every 60s and you want to know _now_ what the status is of switches 4, 5 and 6. But as the device might have several banks (= instance field) you would send PGN 126208, request PGN 127501 Instance 2 (for example). In that case the selection pair would have 1 pair: (field 1, value 2), and the parameters would be 3 long, containing (4, 5, 6). The device could then produce PGN 127501 for instance 2 and not include any switches beyond 6 (as that is the highest field requested).

**Comment by keesverruijt** [comment]
Note that before v4.10.0 (released on 2023-02-10) the analyzer did not produce correct output for PGNs with double variable list arrays, like the read fields PGN.
